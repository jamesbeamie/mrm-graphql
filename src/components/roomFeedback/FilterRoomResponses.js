import React, { Component } from 'react';
import PropTypes from 'prop-types';
import InputRange from 'react-input-range';
import 'react-input-range/src/scss/index.scss';
import { SelectInput } from '../commons';

class FilterRoomResponses extends Component {
  state = {
    sliderValue: {
      min: this.props.sliderSpan.minValue,
      max: this.props.sliderSpan.maxValue,
    },
    roomFilter: '',
  }

  componentDidUpdate({ useFilter: useFilterPrev }) {
    const { useFilter, isActive } = this.props;
    if (useFilterPrev && !useFilter) {
      this.resetFields();
    }
    if (!isActive) {
      document.removeEventListener('click', this.handleOutsideClick);
    } else {
      document.addEventListener('click', this.handleOutsideClick);
    }
  }

  getInitialSliderValue = () => ({
    min: this.props.sliderSpan.minValue,
    max: this.props.sliderSpan.maxValue,
  })

  modal = React.createRef()

  resetFields = () => {
    this.setState({
      sliderValue: this.getInitialSliderValue(),
      roomFilter: '',
    });
  }

  handleSelectChange = ({ target: { value: roomFilter } }) => {
    this.setState({ roomFilter });
  }

  handleSliderChange = (sliderValue) => {
    this.setState({ sliderValue });
  }

  handleFilter = async () => {
    const { filterData, setRoom } = this.props;
    const { roomFilter } = this.state;
    await setRoom(roomFilter);
    filterData();
  }

  clearFilters = () => {
    const { clearFilters } = this.props;
    this.resetFields();
    clearFilters();
  }

  handleOutsideClick = ({ target }) => {
    if (this.modal && this.modal.contains(target)) return;
    this.props.toggleFilterModal(false);
  }

  render() {
    const {
      setResponseCutoff,
      sliderSpan,
      rooms,
      useFilter,
    } = this.props;

    const { sliderValue, roomFilter } = this.state;
    return (
      <div ref={(node) => { this.modal = node; }}>
        <SelectInput
          wrapperClassName="filter-by-room"
          labelText=""
          name="room-name"
          id="add-room-filter"
          isValue
          dynamicValue={roomFilter}
          options={rooms}
          onChange={this.handleSelectChange}
          placeholder="Select Room Name"
        />
        <div>
          <p>Responses:</p>
          <InputRange
            {...sliderSpan}
            draggableTrack
            onChange={this.handleSliderChange}
            onChangeComplete={value => setResponseCutoff(value)}
            value={sliderValue}
          />
        </div>
        <div className="action-buttons">
          {
            useFilter && (
            <button className="btn-secondary" onClick={this.clearFilters}>
              CLEAR FILTERS
            </button>
            )
          }
          <button className="main-button btn-primary" onClick={this.handleFilter}>
            APPLY
          </button>
        </div>
      </div>
    );
  }
}


FilterRoomResponses.propTypes = {
  useFilter: PropTypes.bool.isRequired,
  clearFilters: PropTypes.func.isRequired,
  filterData: PropTypes.func.isRequired,
  setRoom: PropTypes.func.isRequired,
  sliderSpan: PropTypes.shape({
    minValue: PropTypes.number,
    maxValue: PropTypes.number,
  }).isRequired,
  setResponseCutoff: PropTypes.func.isRequired,
  rooms: PropTypes.arrayOf(PropTypes.string).isRequired,
  toggleFilterModal: PropTypes.func.isRequired,
  isActive: PropTypes.bool.isRequired,
};

export default FilterRoomResponses;
