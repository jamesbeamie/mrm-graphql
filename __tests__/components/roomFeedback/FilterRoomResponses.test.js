import React from 'react';
import { shallow } from 'enzyme';
import FilterRoomResponses from '../../../src/components/roomFeedback/FilterRoomResponses';


describe('FilterRoomResponses test', () => {
  let wrapper;
  const props = {
    roomFilter: '',
    setRoom: jest.fn(),
    setResponseCutoff: jest.fn(),
    sliderSpan: { minValue: 0, maxValue: 3 },
    filterData: jest.fn(),
    rooms: ['Tortuga'],
    useFilter: false,
    clearFilters: jest.fn(),
  };

  beforeAll(() => {
    wrapper = shallow(<FilterRoomResponses {...props} />);
  });
  test('Should render correctly', () => {
    expect(wrapper).toMatchSnapshot();
  });

  test('Should select room', () => {
    wrapper.find('SelectInput').dive().find('select').simulate('change', { target: { value: 'Tortuga' } });
    expect(wrapper.state('roomFilter')).toBe('Tortuga');
  });

  test('Should set slider value', () => {
    wrapper.instance().handleSliderChange({ min: 1, max: 4 });
    expect(wrapper.state('sliderValue')).toEqual({ min: 1, max: 4 });
  });

  test('Should call filter handler', () => {
    wrapper.find('.btn-primary').simulate('click');
    expect(props.setRoom).toHaveBeenLastCalledWith('Tortuga');
  });

  test('Should reset fields', () => {
    wrapper.setProps({ useFilter: true });
    wrapper.setProps({ useFilter: false });
    expect(wrapper.state('roomFilter')).toBe('');
    expect(wrapper.state('sliderValue')).toEqual({ min: 0, max: 3 });
  });
});
