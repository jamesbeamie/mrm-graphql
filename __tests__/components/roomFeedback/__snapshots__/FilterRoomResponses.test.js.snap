// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`FilterRoomResponses test Should render correctly 1`] = `
<div>
  <SelectInput
    dynamicValue=""
    id="add-room-filter"
    isValue={true}
    labelText=""
    name="room-name"
    onChange={[Function]}
    onClick={[Function]}
    options={
      Array [
        "Tortuga",
      ]
    }
    placeholder="Select Room Name"
    placeholderValue=""
    required={false}
    selectInputClassName="default-select"
    selected={-1}
    wrapperClassName="filter-by-room"
  />
  <div>
    <p>
      Responses:
    </p>
    <InputRange
      allowSameValues={false}
      classNames={
        Object {
          "activeTrack": "input-range__track input-range__track--active",
          "disabledInputRange": "input-range input-range--disabled",
          "inputRange": "input-range",
          "labelContainer": "input-range__label-container",
          "maxLabel": "input-range__label input-range__label--max",
          "minLabel": "input-range__label input-range__label--min",
          "slider": "input-range__slider",
          "sliderContainer": "input-range__slider-container",
          "track": "input-range__track input-range__track--background",
          "valueLabel": "input-range__label input-range__label--value",
        }
      }
      disabled={false}
      draggableTrack={true}
      maxValue={3}
      minValue={0}
      onChange={[Function]}
      onChangeComplete={[Function]}
      step={1}
      value={
        Object {
          "max": 3,
          "min": 0,
        }
      }
    />
  </div>
  <div
    className="action-buttons"
  >
    <button
      className="main-button btn-primary"
      onClick={[Function]}
    >
      APPLY
    </button>
  </div>
</div>
`;
